### PR TITLE
OAI-38: Include model evaluation

### PR DESCRIPTION
**Pull request by dborowiecki**
TICKET: [OAI-38](https://openimis.atlassian.net/browse/OAI-38)

### CODE REVIEW

File: claim_ai/apps.py [hunk 1]
```diff
@@ -7,22 +7,33 @@
 
 logger = logging.getLogger(__name__)
 
+logger = logging.getLogger(__name__)
```
Comment (dragos-dobre) [inline claim_ai/apps.py]: this is added twice.

File: claim_ai/apps.py [hunk 2]
```diff
@@ -7,22 +7,33 @@
...
+
 MODULE_NAME = 'claim_ai'
 
 DEFAULT_CONFIG = {
     'authentication': [],
     'claim_response_url': 'http://localhost:8000/claim_ai/ClaimResponse',
     'ai_model_file': "",
```
Comment (dragos-dobre) [inline claim_ai/apps.py]: you can add the default ai_model_file. We need to be able to replace this trough docker and .env and to place the files outside docker container.

File: claim_ai/evaluation/ai_model.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                         'ItemFrequency', 'ItemPatCat', 'ItemLevel',
+                         'DateFrom', 'DateTo', 'DateClaimed', 'DOB',
+                         'VisitType', 'HFLevel', 'HFCareType',
+                         'Gender', 'ItemServiceType']
+
+        index = input['SanityCheck'] == 'Clean data'
+        clean_input = input.loc[index, selected_cols].copy()
+        return index, clean_input
+
+    def fill_missing_varaibles(self, clean_input):
```
Comment (dragos-dobre) [inline claim_ai/evaluation/ai_model.py]: fix spelling

File: claim_ai/evaluation/ai_model.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    def evaluate_bundle(self, input_bundle):
+        index, clean_input = self.sanity_check(input_bundle)
+        clean_input = self.fill_missing_varaibles(clean_input)
+        clean_input = self.convert_variables(clean_input)
+        clean_input = self.normalize_input(clean_input)
+        return index, self.predict(clean_input)
+
+    def sanity_check(self, input):
+        date_from = datetime(2016, 5, 15)
+        date_claimed = datetime(2016, 5, 15)
```
Comment (dragos-dobre) [inline claim_ai/evaluation/ai_model.py]: please add date in the configuration with default value 01.01.2016. only one date for both dates here.

File: claim_ai/evaluation/ai_model.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    
+    def convert_variables(self, clean_input):
+        clean_input.loc[:, 'Age'] = (clean_input['DateFrom'] - clean_input['DOB']).dt.days / 365.25
+        # # Drop DOB column as no longer necessary
+        clean_input.drop(['DOB'], axis=1, inplace=True)
+
+        # # Convert to number of days the columns, same date from configuration
+        date_cols = ['DateFrom', 'DateTo', 'DateClaimed']
+        for i in date_cols:
+            clean_input[i] = (clean_input[i] - datetime(2016, 1, 1)).dt.days
```
Comment (dragos-dobre) [inline claim_ai/evaluation/ai_model.py]: use date from configuration.

**Review by dragos-dobre — CHANGES_REQUESTED** [review]

File: claim_ai/apps.py (hunk 1 above)
Comment (dborowiecki) [reply claim_ai/apps.py]: Removed

File: claim_ai/apps.py (hunk 2 above)
Comment (dborowiecki) [reply claim_ai/apps.py]: I added 'joblib_Voting_Model.pkl' as default value

File: claim_ai/evaluation/ai_model.py (hunk 3 above)
Comment (dborowiecki) [reply claim_ai/evaluation/ai_model.py]: Fixed

File: claim_ai/evaluation/ai_model.py (hunk 4 above)
Comment (dborowiecki) [reply claim_ai/evaluation/ai_model.py]: Date changed to be retrieved from the configuration

File: claim_ai/evaluation/ai_model.py (hunk 5 above)
Comment (dborowiecki) [reply claim_ai/evaluation/ai_model.py]: Date from config used

File: claim_ai/apps.py [hunk 6]
```diff
@@ -12,20 +12,31 @@
...
     'authentication': [],
     'claim_response_url': 'http://localhost:8000/claim_ai/ClaimResponse',
-    'ai_model_file': "",
+    'ai_model_file': "joblib_Voting_Model.pkl",
+    'ai_scaler_file': "scaler.pkl",
+    'ai_encoder_file': "Encoder.obj",
     'claim_response_organization': 'openIMIS-Claim-AI',
-    'date_format': '%Y-%m-%d'
+    'date_format': '%Y-%m-%d',
+    'first_date': '2016-05-01'
```
Comment (dragos-dobre) [inline claim_ai/apps.py]: First date is 01.01.2016.

**Review by dragos-dobre — CHANGES_REQUESTED** [review]

File: claim_ai/evaluation/converters/patient.py [hunk 7]
```diff
@@ -46,6 +54,6 @@ def _get_contained_patient_location_code(self, contained_patient):
         return location_extension['valueReference']['identifier']['value']
 
     def _get_contained_patient_group(self, contained_patient):
-        # TODO: This extension is not yet available in fhir api, returns None as default
-        return next((extension['value'] for extension in contained_patient['extension']
-                     if extension['url'].endswith('povertyStatus')) or [], None)
+        group = next(extension for extension in contained_patient['extension']
+                                  if extension['url'].endswith('group'))
+        return group['valueReference']['identifier']['value']
```
Comment (dragos-dobre) [inline claim_ai/evaluation/converters/patient.py]: I mostly saw the references used as "Group/UUID" as short reference. Can we check if reference.reference exists take the UUID from there else search in identifier?

**Review by dragos-dobre — CHANGES_REQUESTED** [review]

File: claim_ai/apps.py (hunk 6 above)
Comment (dborowiecki) [reply claim_ai/apps.py]: Fixed

File: claim_ai/evaluation/converters/patient.py (hunk 7 above)
Comment (dborowiecki) [reply claim_ai/evaluation/converters/patient.py]: Changed for group and also for location

**Review by dragos-dobre — APPROVED** [review]

**Pull request merged**